### PR TITLE
feature: implemented routes API for admin user, allowed user to manage route.

### PR DESCRIPTION
**Pull request by membphis**

### CODE REVIEW

File: conf/nginx.conf [hunk 1]
```diff
@@ -74,6 +74,13 @@ http {
             stub_status;
         }
 
+        location /apisix.com/admin {
```
Comment (moonming) [inline conf/nginx.conf]: better `apisix` instead of `apisix.com`

File: bin/apisix [hunk 2]
```diff
@@ -81,6 +81,15 @@ http {
             stub_status;
         }
 
+        {% if enable_admin then %}
+        location /apisix.com/admin {
```
Comment (moonming) [inline bin/apisix]: better apisix instead of apisix.com

File: conf/nginx.conf (hunk 1 above)
Comment (idevz) [reply conf/nginx.conf]: agree with moonming

File: lua/apisix.lua [hunk 3]
```diff
@@ -41,6 +45,7 @@ function _M.init_worker()
     require("apisix.service").init_worker()
     require("apisix.consumer").init_worker()
     require("apisix.heartbeat").init_worker()
+    require("apisix.admin.init").init_worker()
```
Comment (moonming) [inline lua/apisix.lua]: The admin API will always run whether I disabled the admin API in yaml?

File: t/admin/routes.t [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+no_long_string();
+no_root_location();
+no_shuffle();
+log_level("info");
+
+run_tests;
+
+__DATA__
+
+=== TEST 1: set route(id: 1)
```
Comment (moonming) [inline t/admin/routes.t]: we need more test cases.

File: lua/apisix/core/schema.lua [hunk 5]
```diff
@@ -25,6 +25,14 @@ end
 -- https://github.com/Tencent/rapidjson/blob/master/bin/draft-04/schema
 -- rapidjson not supported `format` in draft-04 yet
 function _M.check(schema, json)
+    if type (schema) ~= "table" then
```
Comment (moonming) [inline lua/apisix/core/schema.lua]: the type of schema and json can be string.

File: lua/apisix/admin/routes.lua [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            "required": [
+                "nodes",
+                "type"
+            ]
+        },
+        "uri": {
+            "type": "string"
+        }
+    },
+    "required": [
```
Comment (moonming) [inline lua/apisix/admin/routes.lua]: all required?

File: lua/apisix/admin/routes.lua [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+
+function _M.post(uri_segs, conf)
+    local resource, id = uri_segs[4], uri_segs[5]
+    if not id then
+        return 200, {error_msg = "missing route id"}
+    end
+
+    local key = "/" .. resource .. "/" .. id
+    local res, err = core.etcd.push(key, conf)
```
Comment (moonming) [inline lua/apisix/admin/routes.lua]: why not check schema here?

File: lua/apisix/admin/routes.lua [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    if not res then
+        core.log.error("failed to get routes[", key, "]: ", err)
```
Comment (moonming) [inline lua/apisix/admin/routes.lua]: `get` should be `post`

File: lua/apisix/admin/routes.lua [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    -- core.log.info("schema: ", core.json.delay_encode(schema_desc))
+    local ok, err = core.schema.check(schema_desc, conf)
+    if not ok then
+        return 400, {error_msg = "invalid configuration: " .. err}
+    end
+
+    local key = "/" .. resource .. "/" .. id
+    local res, err = core.etcd.set(key, conf)
+    if not res then
+        core.log.error("failed to get routes[", key, "]: ", err)
```
Comment (moonming) [inline lua/apisix/admin/routes.lua]: `get` should be `put`

File: lua/apisix/core.lua [hunk 10]
```diff
@@ -12,6 +12,7 @@ return {
     timer    = require("apisix.core.timer"),
     id       = require("apisix.core.id"),
     utils    = require("apisix.core.utils"),
+    etcd     = require("apisix.core.etcd"),
     consumer = require("apisix.consumer"),
     tablepool= require("tablepool"),
```
Comment (moonming) [inline lua/apisix/core.lua]: There should be a space to the left of the `=`

File: lua/apisix.lua (hunk 3 above)
Comment (membphis) [reply lua/apisix.lua]: fixed already

File: lua/apisix/admin/routes.lua (hunk 6 above)
Comment (membphis) [reply lua/apisix/admin/routes.lua]: I mainly used https://www.liquid-technologies.com/online-json-to-schema-converter to generate the schema context, it is hard to write the json schema by hand.

**Comment by membphis** [comment]
@moonming all fixed

File: lua/apisix.lua (hunk 3 above)
Comment (butterfly1924) [reply lua/apisix.lua:48]: @moonming I suggest `enable_admin` to be `enable_admin_api` for clarity.
https://github.com/apache/apisix/blob/f31bef426ef89af950e43058dad2143baec41612/conf/config-default.yaml#L31

**Comment by moonming** [comment]
Welcome create an issue for this, thx

Thanks,
Ming Wen
Twitter: _WenMing


butterfly1938 ***@***.***> 于2022年3月6日周日 00:29写道：

**Pull request merged**